cpu_scheduling_algorithms/non_preemptive_sjf_scheduling.cpp

```diff
@@ -0,0 +1,312 @@
+ * SJF scheduling
+ * @tparam S Data type of Process ID
+ * @tparam T Data type of Arrival time
+ * @tparam E Data type of Burst time
+ * @param input A vector of tuples containing Process ID, Arrival time, and
+ * Burst time
+ * @returns A vector of tuples containing Process ID, Arrival time, Burst time,
+ *         Completion time, Turnaround time, and Waiting time
+ */
+template <typename S, typename T, typename E>
+vector<tuple<S, T, E, double, double, double>> get_final_status(
+    vector<tuple<S, T, E>> input) {
+    // Sort the processes based on Arrival time and then Burst time
+    sort(input.begin(), input.end(), sortcol<S, T, E>);
+
+    // Result vector to hold the final status of each process
+    vector<tuple<S, T, E, double, double, double>> result(input.size());
+    double timeElapsed = 0;
+
+    for (size_t i = 0; i < input.size(); i++) {
+        // Extract Arrival time and Burst time
+        T arrival = get<1>(input[i]);
+        E burst = get<2>(input[i]);
+
+        // If the CPU is idle, move time to the arrival of the next process
+        if (arrival > timeElapsed) {
+            timeElapsed = arrival;
+        }
+
+        // Update timeElapsed by adding the burst time
+        timeElapsed += burst;
+
+        // Calculate Completion time, Turnaround time, and Waiting time
+        double completion = timeElapsed;
+        double turnaround = completion - arrival;
+        double waiting = turnaround - burst;
+
+        // Store the results in the result vector
+        result[i] = make_tuple(get<0>(input[i]), arrival, burst, completion,
+                               turnaround, waiting);
+    }
+
+    return result;
+}
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    // A vector to store the results of all processes across all test cases.
+    vector<tuple<uint32_t, uint32_t, uint32_t, double, double, double>>
+        finalResult;
+
+    for (int i{}; i < 10; i++) {
+        std::random_device rd;  // Seeding
+        std::mt19937 eng(rd());
+        std::uniform_int_distribution<> distr(1, 10);
+
+        uint32_t n = distr(eng);
+        SJF<uint32_t, uint32_t, uint32_t> readyQueue;
+        vector<tuple<uint32_t, uint32_t, uint32_t, double, double, double>>
+            input(n);
+
+        // Generate random arrival and burst times
+        for (uint32_t i{}; i < n; i++) {
+            get<0>(input[i]) = i;
+            get<1>(input[i]) = distr(eng);  // Random arrival time
+            get<2>(input[i]) = distr(eng);  // Random burst time
+        }
+
+        // Print processes before scheduling
+        cout << "Processes before SJF scheduling:" << endl;
+        readyQueue.printResult(input);
+
+        // Add processes to the queue
+        for (uint32_t i{}; i < n; i++) {
+            readyQueue.addProcess(get<0>(input[i]), get<1>(input[i]),
+                                  get<2>(input[i]));
+        }
+
```

Comment:
Please add asset statements instead of printing

Reply:
I have added the assert statement in scheduleforSJF() and commented the printing statements.

@realstealthninja Kindly review the PR.